Library/brew.sh

```diff
@@ -140,12 +173,47 @@ case "$HOMEBREW_COMMAND" in
   --config)    HOMEBREW_COMMAND="config";;
 esac
 
-if [[ -f "$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh" ]] ; then
+if [[ -f "$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh" ]]
+then
   HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND.sh"
-elif [[ -n "$HOMEBREW_DEVELOPER" && -f "$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh" ]] ; then
+elif [[ -n "$HOMEBREW_DEVELOPER" && -f "$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh" ]]
+then
   HOMEBREW_BASH_COMMAND="$HOMEBREW_LIBRARY/Homebrew/dev-cmd/$HOMEBREW_COMMAND.sh"
 fi
 
+if [[ "$HOMEBREW_COMMAND" = "help" || "$1" = "-h" || "$1" = "--help" ]]
+then
+
+  # Display sub command help for `brew help foo` and `brew foo --help`
+  if [[ "$HOMEBREW_ARG_COUNT" -gt 0 ]]
+  then
+    if [[ "$HOMEBREW_COMMAND" = "help" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$1"
+    else
+      HOMEBREW_COMMAND="$HOMEBREW_LIBRARY/Homebrew/cmd/$HOMEBREW_COMMAND"
+    fi
+
+    if [[ -f "$HOMEBREW_COMMAND.sh" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_COMMAND.sh"
+    elif [[ -f "$HOMEBREW_COMMAND.rb" ]]
+    then
+      HOMEBREW_COMMAND="$HOMEBREW_COMMAND.rb"
+    else
+      echo "Unknown command \`$@\`"
+      echo
+      echo "$HOMEBREW_HELP"
+      exit 1
+    fi
+    grep "#:" "$HOMEBREW_COMMAND" | cut -c3- | tr -d \` | sed "s/^\* //"
+    exit 0
+  else
+    echo "$HOMEBREW_HELP"
+    exit 0
+  fi
+fi
```

Comment:
In addition to what Mike said, with what I agree completely:

> Limit us to add actual comment in the head of file.

I don't think this is limiting us at all. Normal comments will continue to be `# <comment text>` (notice the space) instead of starting with `#:`, so there's no ambiguity there.

> Limit us to dynamic process help string. If we were using ruby string, we could dynamic insert say version number or url based on OS.

While this has the danger of creating yet another markup language, there's nothing stopping us from doing some simple replacements on patterns that are very unlikely to appear in the help text, e.g., `@OSX_RELEASE@` gets replaced with `10.11.4` or whatever the current value is.

> It's not easy to access for other purpose, e.g. generating manpage.

It's actually very easy. We already do some processing in `brew man` for conversion to a proper man page and its HTML version. There's nothing stopping us from stitching together the Markdown from all `cmd/*.{rb,sh}` files and adding a header and footer in a preprocessing step.